<a href="https://colab.research.google.com/github/duanfafner/haruandchar/blob/main/Linear_%26_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

SU25:Data Analytics For Business

Homework 1 - Part 1 (Linear Regression)

Dataset: Used Cars


In [6]:
library(readr)
library(dplyr)
library(ggplot2)
library(tidyverse)


In [7]:
data <- read.csv("UsedCars.csv")
summary(data)
head(data, 10)

       Id            Model               Price            Age       
 Min.   :   1.0   Length:1264        Min.   : 5250   Min.   : 1.00  
 1st Qu.: 316.8   Class :character   1st Qu.: 8500   1st Qu.:44.00  
 Median : 632.5   Mode  :character   Median : 9940   Median :61.00  
 Mean   : 632.5                      Mean   :10679   Mean   :56.58  
 3rd Qu.: 948.2                      3rd Qu.:11900   3rd Qu.:70.25  
 Max.   :1264.0                      Max.   :24500   Max.   :80.00  
       KM               HP           Metallic        Automatic     
 Min.   :     1   Min.   : 71.0   Min.   :0.0000   Min.   :0.0000  
 1st Qu.: 40184   1st Qu.: 97.0   1st Qu.:0.0000   1st Qu.:0.0000  
 Median : 60716   Median :110.0   Median :1.0000   Median :0.0000  
 Mean   : 62542   Mean   :104.2   Mean   :0.6756   Mean   :0.0625  
 3rd Qu.: 80474   3rd Qu.:110.0   3rd Qu.:1.0000   3rd Qu.:0.0000  
 Max.   :194545   Max.   :192.0   Max.   :1.0000   Max.   :1.0000  
       CC            Doors           Gear

,Id,Model,Price,Age,KM,HP,Metallic,Automatic,CC,Doors,Gears,Weight
,<int>,<chr>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>,<int>
1,1,TOYOTA Corolla 1800 T SPORT VVT I 2/3-Doors,21500,27,19700,192,0,0,1800,3,5,1185
2,2,TOYOTA Corolla 1.8 VVTL-i T-Sport 3-Drs 2/3-Doors,20950,25,31461,192,0,0,1800,3,6,1185
3,3,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT BNS 2/3-Doors,19950,22,43610,192,0,0,1800,3,6,1185
4,4,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT 2/3-Doors,19600,25,32189,192,0,0,1800,3,6,1185
5,5,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT 2/3-Doors,21500,31,23000,192,1,0,1800,3,6,1185
6,6,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT 2/3-Doors,22500,32,34131,192,1,0,1800,3,6,1185
7,7,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT 2/3-Doors,22000,28,18739,192,0,0,1800,3,6,1185
8,8,TOYOTA Corolla 1.8 16V VVTLI 3DR T SPORT 2/3-Doors,22750,30,34000,192,1,0,1800,3,5,1185
9,9,TOYOTA Corolla 1.6 VVTI Linea Terra Comfort 2/3-Doors,17950,24,21716,110,1,0,1600,3,5,1105


In [8]:
#Run a linear regression of Price on all available variables
model <- lm(Price ~ Age + KM + HP + Metallic + Automatic + CC + Doors + Gears + Weight, data = data)
summary(model)


Call:
lm(formula = Price ~ Age + KM + HP + Metallic + Automatic + CC + 
    Doors + Gears + Weight, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-8117.5  -771.5   -50.0   701.0  6403.5 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -2.883e+03  1.505e+03  -1.916 0.055649 .  
Age         -1.299e+02  2.610e+00 -49.782  < 2e-16 ***
KM          -1.465e-02  1.441e-03 -10.162  < 2e-16 ***
HP           2.540e+01  3.376e+00   7.524 1.01e-13 ***
Metallic    -2.281e+01  7.436e+01  -0.307 0.759059    
Automatic    4.991e+02  1.497e+02   3.334 0.000881 ***
CC          -5.259e-03  8.499e-02  -0.062 0.950667    
Doors        1.315e+01  4.086e+01   0.322 0.747638    
Gears        6.281e+02  1.927e+02   3.259 0.001148 ** 
Weight       1.504e+01  1.300e+00  11.567  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1227 on 1254 degrees of freedom
Multiple R-squared:  0.8649,	Adjusted R-squared

It seems like weight, Gears, whether a car is automatic, HP, KM, Age are affecting the car price. These variables have stronger influence with statistical significance. Overall model p value<0.05, therefore doesnt reject the null hypothesis and is statistical significant. The current list of variables could be used to explain 86% of the model. Automatic has the lagest coefficient estimate.



In [9]:
# Calculate fitted values and residuals
fitted_values <- fitted(model)  # or predict(model)
residuals_values <- residuals(model)  # or model$residuals

# Create a comparison table
results_table <- data.frame(
  Original_Y = data$Price,           # Original y values
  Fitted_Y = fitted_values,          # Fitted ŷ values
  Residuals = residuals_values       # Residuals (y - ŷ)
)

# Display the table
print(results_table)

# Or view first 20 rows for better readability
head(results_table, 20)


     Original_Y  Fitted_Y     Residuals
1         21500 19186.611  2313.3889044
2         20950 19902.383  1047.6167246
3         19950 20114.269  -164.2689291
4         19600 19891.720  -291.7200640
5         21500 19223.830  2276.1696008
6         22500 18930.847  3569.1534518
7         22000 19698.891  2301.1094936
8         22750 18564.507  4185.4929395
9         17950 16239.516  1710.4842222
10        16750 15604.496  1145.5043987
11        16950 14835.242  2114.7583260
12        15950 14786.891  1163.1089591
13        16950 16764.066   185.9341926
14        15950 15438.013   511.9866513
15        16950 15791.437  1158.5627113
16        16250 15755.337   494.6627684
17        15950 16236.493  -286.4928087
18        17495 15887.327  1607.6725044
19        15750 15526.811   223.1894505
20        16950 15639.625  1310.3746671
21        17950 15841.044  2108.9561412
22        12950 15360.746 -2410.7461408
23        15750 15897.604  -147.6035483
24        15950 15331.206   618.7936715


,Original_Y,Fitted_Y,Residuals
,<int>,<dbl>,<dbl>
1,21500,19186.61,2313.3889
2,20950,19902.38,1047.6167
3,19950,20114.27,-164.2689
4,19600,19891.72,-291.7201
5,21500,19223.83,2276.1696
6,22500,18930.85,3569.1535
7,22000,19698.89,2301.1095
8,22750,18564.51,4185.4929
9,17950,16239.52,1710.4842


In [10]:
# prompt: what is the 7th fitted ŷ value and residual, respectively?

cat("7th fitted value:", fitted_values[7], "\n")
cat("7th residual:", residuals_values[7], "\n")

7th fitted value: 19698.89 
7th residual: 2301.109 


In [11]:
#Calculate t-statstics
t_stats <- summary(model)$coefficients[, "t value"]
print(t_stats)

 (Intercept)          Age           KM           HP     Metallic    Automatic 
 -1.91555684 -49.78213520 -10.16168388   7.52407315  -0.30678348   3.33400809 
          CC        Doors        Gears       Weight 
 -0.06188146   0.32182598   3.25899198  11.56676201 


In [12]:
# Step 1: Get the degrees of freedom from the regression model
df <- df.residual(model)
print(paste("Degrees of freedom:", df))


# Step 2: Calculate the critical value for 95% confidence level
# For a two-tailed test at 95% confidence, we need the 97.5th percentile
# (because we have 2.5% in each tail: 2.5% + 95% + 2.5% = 100%)
critical_value <- qt(0.975, df = df)
print(paste("Critical value at 95% confidence level:", critical_value))

# Alternative way to see both positive and negative critical values
cat("Critical values: ±", critical_value, "\n")

[1] "Degrees of freedom: 1254"
[1] "Critical value at 95% confidence level: 1.96185754407177"
Critical values: ± 1.961858 


In [15]:
install.packages("car")

library(car)


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘pbkrtest’


Loading required package: carData


Attaching package: ‘car’


The following object is masked from ‘package:purrr’:

    some


The following object is masked from ‘package:dplyr’:

    recode




In [16]:

# Calculate VIF for all variables
vif_values <- vif(model)
print(vif_values)

# Extract VIF specifically for KM variable
vif_km <- vif_values["KM"]
print(paste("VIF for KM variable:", vif_km))

      Age        KM        HP  Metallic Automatic        CC     Doors     Gears 
 1.914772  1.586299  1.548521  1.017014  1.101810  1.104307  1.269749  1.129618 
   Weight 
 2.096541 
[1] "VIF for KM variable: 1.58629885672294"


Interpreting VIF for multicollinearity:
VIF interpretation guidelines:

VIF = 1: No multicollinearity

1 < VIF < 5: Moderate multicollinearity (generally acceptable)

5 ≤ VIF < 10: High multicollinearity (concerning)

VIF ≥ 10: Severe multicollinearity (problematic)


What VIF means:

VIF = 5 means the variance of the coefficient is 5 times larger due to multicollinearity
VIF = 10 means the standard error is √10 ≈ 3.16 times larger than it would be without multicollinearity

In [18]:
# Regress Weight on all other explanatory variables (excluding Price)
weight_model <- lm(Weight ~ Age + KM + HP + Metallic + Automatic + CC + Doors + Gears, data = data)
summary(weight_model)


Call:
lm(formula = Weight ~ Age + KM + HP + Metallic + Automatic + 
    CC + Doors + Gears, data = data)

Residuals:
   Min     1Q Median     3Q    Max 
-62.22 -12.80  -0.52   7.99 527.97 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  8.739e+02  2.143e+01  40.784   <2e-16 ***
Age         -7.322e-01  5.278e-02 -13.873   <2e-16 ***
KM          -1.229e-06  3.130e-05  -0.039   0.9687    
HP           1.210e+00  6.486e-02  18.651   <2e-16 ***
Metallic    -1.061e+00  1.614e+00  -0.657   0.5112    
Automatic    2.953e+01  3.142e+00   9.399   <2e-16 ***
CC           3.710e-03  1.842e-03   2.014   0.0443 *  
Doors        1.174e+01  8.231e-01  14.261   <2e-16 ***
Gears        9.766e+00  4.176e+00   2.339   0.0195 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 26.65 on 1255 degrees of freedom
Multiple R-squared:  0.523,	Adjusted R-squared:   0.52 
F-statistic:   172 on 8 and 1255 DF,  p-value: < 2.2e-16


In [19]:

# Get the R-squared value
r_squared_weight <- summary(weight_model)$r.squared
print(paste("R-squared for Weight regression:", r_squared_weight))
vif_weight <- 1 / (1 - r_squared_weight)
print(paste("VIF for Weight:", vif_weight))

[1] "R-squared for Weight regression: 0.523023869724916"
[1] "VIF for Weight: 2.09654097244505"


It shows how well the other variables predict Weight:

R² close to 0 → Weight is not well predicted by other variables → low multicollinearity

R² close to 1 → Weight is highly predicted by other variables → high multicollinearity

In [20]:
model2 <- lm(Price ~ Age + KM + HP  + Automatic  + Gears + Weight, data = data)
summary(model)
summary(model2)


Call:
lm(formula = Price ~ Age + KM + HP + Metallic + Automatic + CC + 
    Doors + Gears + Weight, data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-8117.5  -771.5   -50.0   701.0  6403.5 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -2.883e+03  1.505e+03  -1.916 0.055649 .  
Age         -1.299e+02  2.610e+00 -49.782  < 2e-16 ***
KM          -1.465e-02  1.441e-03 -10.162  < 2e-16 ***
HP           2.540e+01  3.376e+00   7.524 1.01e-13 ***
Metallic    -2.281e+01  7.436e+01  -0.307 0.759059    
Automatic    4.991e+02  1.497e+02   3.334 0.000881 ***
CC          -5.259e-03  8.499e-02  -0.062 0.950667    
Doors        1.315e+01  4.086e+01   0.322 0.747638    
Gears        6.281e+02  1.927e+02   3.259 0.001148 ** 
Weight       1.504e+01  1.300e+00  11.567  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1227 on 1254 degrees of freedom
Multiple R-squared:  0.8649,	Adjusted R-squared


Call:
lm(formula = Price ~ Age + KM + HP + Automatic + Gears + Weight, 
    data = data)

Residuals:
    Min      1Q  Median      3Q     Max 
-8193.8  -772.8   -47.9   703.3  6402.3 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -2.930e+03  1.498e+03  -1.956 0.050666 .  
Age         -1.299e+02  2.600e+00 -49.942  < 2e-16 ***
KM          -1.463e-02  1.439e-03 -10.170  < 2e-16 ***
HP           2.527e+01  3.294e+00   7.671  3.4e-14 ***
Automatic    4.930e+02  1.476e+02   3.341 0.000861 ***
Gears        6.139e+02  1.876e+02   3.272 0.001098 ** 
Weight       1.519e+01  1.202e+00  12.637  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1226 on 1257 degrees of freedom
Multiple R-squared:  0.8648,	Adjusted R-squared:  0.8642 
F-statistic:  1341 on 6 and 1257 DF,  p-value: < 2.2e-16


1. Minimal Impact on Explanatory Power:

The R-squared decreased by only 0.0001 (0.8649 → 0.8648) when removing 3 variables
This tiny decrease indicates the removed variables (Metallic, CC, Doors) contributed almost nothing to explaining price variation

2. Improved Model Efficiency:

The Adjusted R-squared actually increased from 0.8639 to 0.8642


3. What This Implies:
Model Parsimony: The reduced model is more efficient - it explains virtually the same amount of variation (86.48% vs 86.49%) using fewer variables


Better Statistical Properties: The reduced model has:

Higher degrees of freedom (1257 vs 1254)
Slightly better adjusted R²
More focused on truly significant predictors

4. Practical Conclusion:
The reduced model is preferable because it achieves the same predictive power with greater simplicity. This follows the principle of Occam's razor - when two models perform similarly, choose the simpler one. The fact that adjusted R² improved confirms that removing the insignificant variables enhanced the model's quality by eliminating unnecessary complexity without sacrificing explanatory power.


SU25:Data Analytics For Business

Homework 2 - Part 2 (Binary Response Model)

Dataset: Loan


In [23]:
data2 <- read.csv("Loan.csv")

summary(data2)
head(data2, 10)
##CCAvg= Average spending on credit cards per month in $1000
##Income = Avg income of customers in $1000
##Loan = whether or not the customer accepted hte persoanl loan offer 1- yes 0-no
##Education level: 1: Undergrad; 2: Graduate; 3: Advanced/Professional
##Family = Family size of the customer

      Loan           Income           Family          CCAvg       
 Min.   :0.000   Min.   :  8.00   Min.   :1.000   Min.   : 0.000  
 1st Qu.:0.000   1st Qu.: 39.00   1st Qu.:1.000   1st Qu.: 0.700  
 Median :0.000   Median : 64.00   Median :2.000   Median : 1.500  
 Mean   :0.096   Mean   : 73.77   Mean   :2.396   Mean   : 1.938  
 3rd Qu.:0.000   3rd Qu.: 98.00   3rd Qu.:3.000   3rd Qu.: 2.500  
 Max.   :1.000   Max.   :224.00   Max.   :4.000   Max.   :10.000  
   Education    
 Min.   :1.000  
 1st Qu.:1.000  
 Median :2.000  
 Mean   :1.881  
 3rd Qu.:3.000  
 Max.   :3.000  

,Loan,Income,Family,CCAvg,Education
,<int>,<int>,<int>,<dbl>,<int>
1,0,49,4,1.6,1
2,0,34,3,1.5,1
3,0,11,1,1.0,1
4,0,100,1,2.7,2
5,0,45,4,1.0,2
6,0,29,4,0.4,2
7,0,72,2,1.5,2
8,0,22,1,0.3,3
9,0,81,3,0.6,2


In [24]:
data2$Education <- as.factor(data2$Education)

In [25]:
# Regress Loan (binary outcome) on all other variables
lpm_model <- lm(Loan ~ Income + Family + CCAvg + Education, data = data2)

# Show the summary of regression results
summary(lpm_model)


Call:
lm(formula = Loan ~ Income + Family + CCAvg + Education, data = data2)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.56354 -0.14730 -0.03822  0.06978  1.05386 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -3.455e-01  1.127e-02 -30.653  < 2e-16 ***
Income       3.367e-03  9.799e-05  34.364  < 2e-16 ***
Family       3.160e-02  3.010e-03  10.499  < 2e-16 ***
CCAvg        1.373e-02  2.538e-03   5.412 6.52e-08 ***
Education2   1.517e-01  8.473e-03  17.908  < 2e-16 ***
Education3   1.605e-01  8.229e-03  19.511  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2393 on 4994 degrees of freedom
Multiple R-squared:  0.3409,	Adjusted R-squared:  0.3402 
F-statistic: 516.6 on 5 and 4994 DF,  p-value: < 2.2e-16


In [26]:
# Get fitted values from the linear probability model
fitted_values <- fitted(lpm_model)  # or predict(lmp_model)

# Find customers with fitted values > 1 or < 0
customers_over_1 <- which(fitted_values > 1)
customers_under_0 <- which(fitted_values < 0)

print(customers_over_1)
length(customers_over_1)
length(customers_under_0)


named integer(0)


[1] 0

[1] 1508

In [27]:
# Get fitted values from the linear probability model
fitted_values <- fitted(lpm_model)  # Make sure you're using the correct model name

# Find the smallest fitted value
min_fitted <- min(fitted_values)
print(paste("Smallest fitted y value:", min_fitted))

[1] "Smallest fitted y value: -0.285626945615051"


In [28]:
# Run Logit model using glm() function
logit_model <- glm(Loan ~ Income + Family + CCAvg + Education,
                   data = data2,
                   family = binomial(link = "logit"))

# Show the summary of estimation results
summary(logit_model)


Call:
glm(formula = Loan ~ Income + Family + CCAvg + Education, family = binomial(link = "logit"), 
    data = data2)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -13.177833   0.517777 -25.451  < 2e-16 ***
Income        0.059791   0.002687  22.255  < 2e-16 ***
Family        0.587079   0.071275   8.237  < 2e-16 ***
CCAvg         0.162679   0.040505   4.016 5.91e-05 ***
Education2    3.910609   0.251037  15.578  < 2e-16 ***
Education3    3.933173   0.244329  16.098  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3162.0  on 4999  degrees of freedom
Residual deviance: 1334.8  on 4994  degrees of freedom
AIC: 1346.8

Number of Fisher Scoring iterations: 8


In [29]:
# Get fitted probabilities from logit model
fitted_probabilities <- predict(logit_model, type="response")

# Calculate threshold (fraction of success in original data)
threshold <- mean(data2$Loan)

# Create binary predictions
predicted_loan <- ifelse(fitted_probabilities > threshold, 1, 0)

# Create confusion matrix
confusion_matrix <- table(Actual = data2$Loan, Predicted = predicted_loan)
print(confusion_matrix)

# Calculate overall PCP
total_correct <- sum(diag(confusion_matrix))
total_observations <- sum(confusion_matrix)
overall_pcp <- (total_correct / total_observations) * 100

cat("Overall PCP:", round(overall_pcp, 2), "%")


      Predicted
Actual    0    1
     0 4001  519
     1   61  419
Overall PCP: 88.4 %

In [30]:
# prompt: what are the correct PCP values for y=1 and y=0, respectively?

# Correctly Classified for y=1 (Actual 1, Predicted 1)
pcp_y1 <- confusion_matrix[2, 2] / sum(confusion_matrix[2, ]) * 100

# Correctly Classified for y=0 (Actual 0, Predicted 0)
pcp_y0 <- confusion_matrix[1, 1] / sum(confusion_matrix[1, ]) * 100

cat("\nPCP for y=1:", round(pcp_y1, 2), "%")
cat("\nPCP for y=0:", round(pcp_y0, 2), "%")


PCP for y=1: 87.29 %
PCP for y=0: 88.52 %

In [31]:
# prompt: Calculate the predicted probability of success according to the Logit model estimation
# results. Evaluate the probability at such values of the X variables: {Income, Family, CCAvg}
# equal their mean values in the original data and Education=”2”.
# • First calculate the probability using the defining formula as explained in class. You may
# do the calculation in R, but be explicit about the exact formula used and substitute in
# the exact numbers.
# • Then use the predict(…, type=”response”) function to obtain the predicted probability
# of success. The value should be the same as your own calculation result above.

# Define the values for the calculation
income_mean <- mean(data2$Income)
family_mean <- mean(data2$Family)
ccavg_mean <- mean(data2$CCAvg)
education_level <- "2" # Corresponds to the factor level "2"

# Get the coefficients from the logit model
coefficients <- coef(logit_model)
intercept <- coefficients["(Intercept)"]
income_coef <- coefficients["Income"]
family_coef <- coefficients["Family"]
ccavg_coef <- coefficients["CCAvg"]
education_coef <- coefficients[paste0("Education", education_level)] # Get coefficient for the specific education level

# Calculate the linear predictor (Z) using the formula
Z <- intercept + (income_mean * income_coef) + (family_mean * family_coef) + (ccavg_mean * ccavg_coef) + education_coef

# Calculate the predicted probability using the logistic function: P = 1 / (1 + exp(-Z))
predicted_prob_manual <- 1 / (1 + exp(-Z))

cat("Manual Calculation of Predicted Probability:\n")
cat("  Formula: P = 1 / (1 + exp(-Z))\n")
cat("  Where Z = (Intercept) + (Income * Coef_Income) + (Family * Coef_Family) + (CCAvg * Coef_CCAvg) + (Education_2 * Coef_Education_2)\n")
cat("  Substituting values:\n")
cat("    Z = ", intercept, " + (", income_mean, " * ", income_coef, ") + (", family_mean, " * ", family_coef, ") + (", ccavg_mean, " * ", ccavg_coef, ") + (1 * ", education_coef, ")\n", sep="")
cat("    Z = ", Z, "\n")
cat("  Predicted Probability = 1 / (1 + exp(-", Z, ")) = ", predicted_prob_manual, "\n\n", sep="")


# Use the predict function to obtain the predicted probability
# Create a new data frame with the specific values for prediction
newdata <- data.frame(
  Income = income_mean,
  Family = family_mean,
  CCAvg = ccavg_mean,
  Education = as.factor(education_level)
)

predicted_prob_predict <- predict(logit_model, newdata = newdata, type = "response")

cat("Predicted Probability using predict() function:\n")
cat("  Predicted Probability = ", predicted_prob_predict, "\n")

# Verify if the results are the same (within a small tolerance for floating point errors)
tolerance <- 1e-6
if (abs(predicted_prob_manual - predicted_prob_predict) < tolerance) {
  cat("\nManual calculation and predict() function results match.\n")
} else {
  cat("\nWarning: Manual calculation and predict() function results do not exactly match.\n")
}

Manual Calculation of Predicted Probability:
  Formula: P = 1 / (1 + exp(-Z))
  Where Z = (Intercept) + (Income * Coef_Income) + (Family * Coef_Family) + (CCAvg * Coef_CCAvg) + (Education_2 * Coef_Education_2)
  Substituting values:
    Z = -13.17783 + (73.7742 * 0.05979075) + (2.3964 * 0.5870788) + (1.937938 * 0.1626791) + (1 * 3.910609)
    Z =  -3.134071 
  Predicted Probability = 1 / (1 + exp(--3.134071)) = 0.04172353

Predicted Probability using predict() function:
  Predicted Probability =  0.04172353 

Manual calculation and predict() function results match.


In [32]:
summary(lpm_model)
summary(logit_model)


Call:
lm(formula = Loan ~ Income + Family + CCAvg + Education, data = data2)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.56354 -0.14730 -0.03822  0.06978  1.05386 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -3.455e-01  1.127e-02 -30.653  < 2e-16 ***
Income       3.367e-03  9.799e-05  34.364  < 2e-16 ***
Family       3.160e-02  3.010e-03  10.499  < 2e-16 ***
CCAvg        1.373e-02  2.538e-03   5.412 6.52e-08 ***
Education2   1.517e-01  8.473e-03  17.908  < 2e-16 ***
Education3   1.605e-01  8.229e-03  19.511  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2393 on 4994 degrees of freedom
Multiple R-squared:  0.3409,	Adjusted R-squared:  0.3402 
F-statistic: 516.6 on 5 and 4994 DF,  p-value: < 2.2e-16



Call:
glm(formula = Loan ~ Income + Family + CCAvg + Education, family = binomial(link = "logit"), 
    data = data2)

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -13.177833   0.517777 -25.451  < 2e-16 ***
Income        0.059791   0.002687  22.255  < 2e-16 ***
Family        0.587079   0.071275   8.237  < 2e-16 ***
CCAvg         0.162679   0.040505   4.016 5.91e-05 ***
Education2    3.910609   0.251037  15.578  < 2e-16 ***
Education3    3.933173   0.244329  16.098  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3162.0  on 4999  degrees of freedom
Residual deviance: 1334.8  on 4994  degrees of freedom
AIC: 1346.8

Number of Fisher Scoring iterations: 8


In [33]:
# prompt: Calculate the partial effects of all independent variables based on the Logit model. Again,
# evaluate the probability at such values of the independent variables: {Income, Family,
# CCAvg} equal their mean values in the original data and Education=”2”. Also calculate the
# partial effects based on the linear probability model (Hint: simply the 𝛽𝛽 coefficients). Co-list
# and compare the partial effects from the two models. Are they comparable

# --- Logit Model Partial Effects ---

# Calculate the predicted probability at the specified values (already done above)
# predicted_prob_at_means_edu2 <- predicted_prob_predict # This is the probability P(Y=1 | X=x_bar, Edu=2)

# Get the coefficients from the Logit model
logit_coefs <- coef(logit_model)

# Calculate partial effects for continuous variables (Income, Family, CCAvg)
# PE = beta_k * P(Y=1 | X) * (1 - P(Y=1 | X))
pe_income_logit <- logit_coefs["Income"] * predicted_prob_predict * (1 - predicted_prob_predict)
pe_family_logit <- logit_coefs["Family"] * predicted_prob_predict * (1 - predicted_prob_predict)
pe_ccavg_logit <- logit_coefs["CCAvg"] * predicted_prob_predict * (1 - predicted_prob_predict)

# Calculate partial effect for dummy variable (Education, assuming baseline is Education=1)
# The PE for moving from Education=1 to Education=2 (at the specified values)
# is P(Y=1 | Income, Family, CCAvg, Edu=2) - P(Y=1 | Income, Family, CCAvg, Edu=1)

# Probability at Edu=2 (already calculated as predicted_prob_predict)
prob_edu2 <- predicted_prob_predict

# Probability at Edu=1 (keeping other variables at specified means)
newdata_edu1 <- data.frame(
  Income = income_mean,
  Family = family_mean,
  CCAvg = ccavg_mean,
  Education = as.factor("1") # Set Education to level 1
)
prob_edu1 <- predict(logit_model, newdata = newdata_edu1, type = "response")

# Partial effect for Education (moving from 1 to 2)
pe_edu_logit <- prob_edu2 - prob_edu1

cat("\n--- Logit Model Partial Effects at Mean Income/Family/CCAvg and Education='2' ---\n")
cat("Partial Effect of Income:  ", pe_income_logit, "\n")
cat("Partial Effect of Family:  ", pe_family_logit, "\n")
cat("Partial Effect of CCAvg:   ", pe_ccavg_logit, "\n")
cat("Partial Effect of Education (1->2): ", pe_edu_logit, "\n")


# --- Linear Probability Model Partial Effects ---

# Get the coefficients from the LPM model
lpm_coefs <- coef(lpm_model)

# Partial effects for LPM are simply the coefficients
pe_income_lpm <- lpm_coefs["Income"]
pe_family_lpm <- lpm_coefs["Family"]
pe_ccavg_lpm <- lpm_coefs["CCAvg"]
# For the dummy variable Education, the coefficient represents the change
# in probability when moving from the base level (Education=1) to the specified level (Education=2).
pe_edu_lpm <- lpm_coefs["Education2"]

cat("\n--- Linear Probability Model Partial Effects (Coefficients) ---\n")
cat("Partial Effect of Income:  ", pe_income_lpm, "\n")
cat("Partial Effect of Family:  ", pe_family_lpm, "\n")
cat("Partial Effect of CCAvg:   ", pe_ccavg_lpm, "\n")
cat("Partial Effect of Education (1->2): ", pe_edu_lpm, "\n")


# --- Comparison ---

cat("\n--- Comparison of Partial Effects (Logit vs LPM) ---\n")
comparison_table <- data.frame(
  Variable = c("Income", "Family", "CCAvg", "Education (1->2)"),
  Logit_PE = c(pe_income_logit, pe_family_logit, pe_ccavg_logit, pe_edu_logit),
  LPM_PE = c(pe_income_lpm, pe_family_lpm, pe_ccavg_lpm, pe_edu_lpm)
)
print(comparison_table)

cat("\nAre the partial effects comparable?\n")
cat("Generally, yes, the signs and relative magnitudes should be broadly similar.\n")
cat("However, the *exact* values will differ because:\n")
cat("1. LPM assumes a linear relationship between predictors and probability,\n")
cat("   which is incorrect for a binary outcome (probabilities must be between 0 and 1).\n")
cat("2. Logit accounts for the non-linear S-shaped relationship of probability.\n")
cat("   Logit partial effects for continuous variables depend on the specific values of the predictors (P(Y=1|X) term).\n")
cat("In this case, let's look at the table:\n")
# Check the differences
diff_income <- abs(pe_income_logit - pe_income_lpm)
diff_family <- abs(pe_family_logit - pe_family_lpm)
diff_ccavg <- abs(pe_ccavg_logit - pe_ccavg_lpm)
diff_edu <- abs(pe_edu_logit - pe_edu_lpm)

cat(paste0("- Income PE difference: ", round(diff_income, 4), "\n"))
cat(paste0("- Family PE difference: ", round(diff_family, 4), "\n"))
cat(paste0("- CCAvg PE difference: ", round(diff_ccavg, 4), "\n"))
cat(paste0("- Education PE difference: ", round(diff_edu, 4), "\n"))

cat("The differences suggest that the LPM likely over/underestimates the effects compared to the Logit model, depending on the variable and the point of evaluation.\n")
cat("LPM is simpler but can produce probabilities outside [0,1] and has constant partial effects (the coefficients).\n")
cat("Logit is more theoretically sound for binary outcomes and has partial effects that vary with the predictor values.\n")
cat("Based on these differences, while broadly aligned, the Logit model provides a more accurate estimate of the change in probability for a unit change in a predictor at this specific point in the data (means + Edu=2).\n")
cat("For continuous variables, the Logit PE is the slope of the probability curve at that point, while the LPM PE is the average slope assuming linearity.\n")
cat("For dummy variables, the Logit PE is the discrete change in probability, while the LPM PE is simply the coefficient difference.\n")
cat("The comparison table shows that the numerical values are different, highlighting the non-linearity captured by the Logit model compared to the linear approximation of the LPM.\n")


--- Logit Model Partial Effects at Mean Income/Family/CCAvg and Education='2' ---
Partial Effect of Income:   0.002390594 
Partial Effect of Family:   0.02347298 
Partial Effect of CCAvg:    0.006504346 
Partial Effect of Education (1->2):  0.04085225 

--- Linear Probability Model Partial Effects (Coefficients) ---
Partial Effect of Income:   0.003367256 
Partial Effect of Family:   0.03160246 
Partial Effect of CCAvg:    0.01373463 
Partial Effect of Education (1->2):  0.1517432 

--- Comparison of Partial Effects (Logit vs LPM) ---
               Variable    Logit_PE      LPM_PE
Income           Income 0.002390594 0.003367256
Family           Family 0.023472981 0.031602457
CCAvg             CCAvg 0.006504346 0.013734629
1      Education (1->2) 0.040852250 0.151743228

Are the partial effects comparable?
Generally, yes, the signs and relative magnitudes should be broadly similar.
However, the *exact* values will differ because:
1. LPM assumes a linear relationship between predictors

In [34]:
# prompt: "They are not directly comparable because the partial effects from the linear probability model equal the coefficients, but the partial effects from the logit model are transformations of the coefficients." is this statement true

# Yes, that statement is true.
# The partial effects in a Linear Probability Model (LPM) are indeed equal to the coefficients of the explanatory variables. A unit change in a predictor directly changes the predicted probability by its coefficient value.
# In contrast, the partial effects (or marginal effects) in a Logit model are *not* simply the coefficients. They are calculated based on the logistic function's derivative (for continuous variables) or as the difference in predicted probabilities (for discrete/dummy variables) at a specific point in the data (e.g., at the mean of the predictors). The formula for the partial effect of a continuous variable X_k in a Logit model is beta_k * Lambda(beta'X), where beta_k is the coefficient for X_k, and Lambda(.) is the PDF of the logistic distribution evaluated at the linear predictor beta'X.
# This non-linear transformation means the partial effects in a Logit model vary depending on the values of *all* the predictors, while they are constant in an LPM. Therefore, the numerical values of partial effects are generally different between the two models, making the statement accurate.

In [36]:
# prompt: Everything else equal, customers with an advanced/professional degree have approximately 15.7% higher probability of accepting the loan offer compared to those with an undergraduate degree. is this statement true

# Calculate the difference in probability between Education=3 and Education=1 using the Logit model
# Probability at Edu=3 (keeping other variables at specified means - Income, Family, CCAvg)
newdata_edu3 <- data.frame(
  Income = income_mean,
  Family = family_mean,
  CCAvg = ccavg_mean,
  Education = as.factor("3") # Set Education to level 3
)
prob_edu3 <- predict(logit_model, newdata = newdata_edu3, type = "response")

# Probability at Edu=1 (keeping other variables at specified means)
# This is the same as prob_edu1 calculated earlier for the 1->2 comparison
# newdata_edu1 is already defined

prob_edu1 <- predict(logit_model, newdata = newdata_edu1, type = "response")


# Calculate the discrete change in probability when moving from Education=1 to Education=3
pe_edu1_to_edu3_logit <- prob_edu3 - prob_edu1

cat("\n--- Calculation for statement verification ---\n")
cat("Predicted Probability at Mean Income/Family/CCAvg and Education='1': ", prob_edu1, "\n")
cat("Predicted Probability at Mean Income/Family/CCAvg and Education='3': ", prob_edu3, "\n")
cat("Difference in Probability (Edu=3 vs Edu=1): ", pe_edu1_to_edu3_logit, "\n")
cat("Percentage Difference: ", pe_edu1_to_edu3_logit * 100, "%\n")

# The statement says the probability is approximately 15.7% higher.
# Let's check if the calculated difference is close to 0.157

if (abs(pe_edu1_to_edu3_logit - 0.157) < 0.01) { # Using a tolerance of 1 percentage point (0.01)
  cat("Based on the logit model calculated at the mean of other variables, the statement that customers with an advanced/professional degree have approximately 15.7% higher probability of accepting the loan offer compared to those with an undergraduate degree is TRUE (the difference is approximately ", round(pe_edu1_to_edu3_logit * 100, 1), "%).", sep="")
} else {
  cat("Based on the logit model calculated at the mean of other variables, the statement is FALSE. The calculated difference is ", round(pe_edu1_to_edu3_logit * 100, 1), "%, not approximately 15.7%.", sep="")
}

# Additional check using LPM (though Logit is preferred)
pe_edu1_to_edu3_lpm <- lpm_coefs["Education3"]

cat("\n--- Check using LPM (for comparison) ---\n")
cat("LPM Coefficient for Education3 (Edu=3 vs Edu=1 baseline): ", pe_edu1_to_edu3_lpm, "\n")
cat("Percentage difference according to LPM: ", pe_edu1_to_edu3_lpm * 100, "%\n")

# The LPM coefficient is also close to 0.157, reinforcing that both models
# suggest a similar magnitude of effect for this discrete change.
```

ERROR: Error in parse(text = input): attempt to use zero-length variable name
